client/jetpack-cloud/controller.ts

```diff
@@ -0,0 +1,217 @@
+/**
+ * External dependencies
+ */
+import page from 'page';
+
+/**
+ * Internal dependencies
+ */
+import { addQueryArgs, getSiteFragment, sectionify } from 'calypso/lib/route';
+import {
+	redirectToPrimary,
+	updateRecentSitesPreferences,
+	renderEmptySites,
+	renderNoVisibleSites,
+	recordNoSitesPageView,
+	recordNoVisibleSitesPageView,
+	showMissingPrimaryError,
+} from 'calypso/my-sites/controller';
+import { getCurrentUser } from 'calypso/state/current-user/selectors';
+import getPrimarySiteId from 'calypso/state/selectors/get-primary-site-id';
+import getSiteId from 'calypso/state/selectors/get-site-id';
+import { requestSite } from 'calypso/state/sites/actions';
+import { getSiteSlug } from 'calypso/state/sites/selectors';
+import { setSelectedSiteId, setAllSitesSelected } from 'calypso/state/ui/actions';
+
+/**
+ * Type dependencies
+ */
+import type { UserData } from 'calypso/lib/user/user';
+import type PageJS from 'page';
+
+/**
+ * Parse site slug from path. If no slug is detected but a `site` query parameter
+ * exists, a redirection to `/:path/:site/` occurs.
+ *
+ * @param {PageJS.Context} context Route context
+ * @returns {string} Site slug
+ */
+const parseSiteFragment = ( context: PageJS.Context ): string | undefined => {
+	const siteFragment = context.params.site || getSiteFragment( context.path );
+
+	if ( siteFragment ) {
+		return siteFragment;
+	}
+
+	const { query: queryParams, pathname } = context;
+	const { site: siteQuery } = queryParams;
+
+	if ( siteQuery ) {
+		page.redirect(
+			addQueryArgs(
+				{
+					...queryParams,
+					site: undefined,
+				},
+				`${ pathname }/${ siteQuery }`
+			)
+		);
+	}
+};
+
+/**
+ * Fetch site data.
+ *
+ * @param {PageJS.Context} context Route context
+ * @param {number|string} siteId Site id
+ * @returns {Promise} Promise that resolves with the site id and slug
+ */
+const fetchSite = (
+	context: PageJS.Context,
+	siteId: number | string
```

Comment:
For precision, we might want to update this argument to something like `siteIdOrSlug`